server/middleware/setUpAuthentication.ts

```diff
@@ -21,12 +22,35 @@ export default function setUpAuth(): Router {
 
   router.get('/sign-in', passport.authenticate('oauth2'))
 
-  router.get('/sign-in/callback', (req, res, next) =>
-    passport.authenticate('oauth2', {
-      successReturnToOrRedirect: req.session.returnTo || '/',
-      failureRedirect: '/autherror',
-    })(req, res, next),
-  )
+  router.get('/sign-in/callback', (req, res, next) => {
+    const authCallback: passport.AuthenticateCallback = (err, user, info) => {
+      if (err) {
+        return res.redirect('/autherror')
+      }
+      if (!user) {
+        const { message } = info as { message: string }
+        if (info && message === 'Unable to verify authorization request state.') {
+          // failure to due authorisation state not being there on return, so retry
+          logger.info('Retrying auth callback as no state found')
+          return res.redirect('/')
+        }
+        logger.info(`Auth failure due to ${JSON.stringify(info)}`)
+        return res.redirect('/autherror')
+      }
+      const { returnTo } = req.session
```

Comment:
Same here, if we're confident `returnTo` always exists in session then this is fine but if there's a chance it won't then we're better of checking it exists before destructuring.

Reply:
I can't be sure so let's guard against it to be safe 👍